### PR TITLE
feature/anton_apr1

### PR DESCRIPTION
**Pull request by anton-anufriev**
Added Log, LogEntry, CommandLogEntry
Change CommandMessage. Added AppendRequestHandler for all states and overrode logic for Candidate state.
De-inherited DirectCommand from Message.
Added replayEvent flag to Transition.

### CODE REVIEW

File: src/main/java/org/tools4j/hoverraft/event/AppendRequestHandler.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+import org.tools4j.hoverraft.message.AppendRequest;
+import org.tools4j.hoverraft.server.ServerContext;
+import org.tools4j.hoverraft.state.*;
+
+import java.util.Objects;
+
+public class AppendRequestHandler {
+    private final PersistentState persistentState;
+    private final VolatileState volatileState;
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/event/AppendRequestHandler.java:11]: ~~you don't need those member variables cause they are passed to the onAppendRequest method via ServerContext~~

File: src/main/java/org/tools4j/hoverraft/event/AppendRequestHandler.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+        return transition;
+    }
+
+    private boolean appendToLog(final ServerContext serverContext, final AppendRequest appendRequest) {
+        final LogEntry prevLogEntry = appendRequest.prevLogEntry();
+        final CommandLogEntry newCommandLogEntry = appendRequest.commandLogEntry();
+        final CommandLog commandLog = persistentState.commandLog();
+
+        CommandLog.CONTAINMENT containment =  commandLog.contains(prevLogEntry);
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/event/AppendRequestHandler.java]: final

File: src/main/java/org/tools4j/hoverraft/message/MessageType.java [hunk 3]
```diff
@@ -35,7 +36,7 @@ public Message create(final MessageFactory factory) {
             return factory.voteRequest();
         }
         @Override
-        public DirectMessage create(final DirectMessageFactory factory) {
+        public AbstractDirectMessage create(final DirectMessageFactory factory) {
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/message/MessageType.java]: I don't like if abstract types are exposed anywhere

File: src/main/java/org/tools4j/hoverraft/message/direct/DirectAppendRequest.java [hunk 4]
```diff
@@ -93,39 +107,41 @@ public AppendRequest leaderId(final int leaderId) {
...
+        return directCommandLogEntry;
     }
 
     @Override
-    public AppendRequest leaderCommit(final long leaderCommit) {
-        writeBuffer.putLong(offset + LEADER_COMMIT_OFF, leaderCommit);
-        return this;
+    public void wrap(DirectBuffer buffer, int offset) {
+        super.wrap(buffer, offset);
+        directCommandLogEntry.wrap(buffer, offset + COMMAND_LOG_ENTRY_OFF);
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/message/direct/DirectAppendRequest.java:109]: nice!

File: src/main/java/org/tools4j/hoverraft/message/direct/DirectAppendRequest.java [hunk 5]
```diff
@@ -93,39 +107,41 @@ public AppendRequest leaderId(final int leaderId) {
...
-        return command;
+    public void wrap(MutableDirectBuffer buffer, int offset) {
+        super.wrap(buffer, offset);
+        directCommandLogEntry.wrap(buffer, offset + COMMAND_LOG_ENTRY_OFF);
     }
 
+    @Override
+    public void unwrap() {
+        super.unwrap();
+        directCommandLogEntry.unwrap();
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/message/direct/DirectAppendRequest.java:121]: should be opposite unwrapping order imo

File: src/main/java/org/tools4j/hoverraft/message/direct/DirectMessageFactory.java [hunk 6]
```diff
@@ -84,11 +84,11 @@ public static DirectMessageFactory createForWriting(final MutableDirectBuffer bu
         return factory;
     }
 
-    public DirectMessage wrapForReading(final DirectBuffer directBuffer, final int offset) {
+    public AbstractDirectMessage wrapForReading(final DirectBuffer directBuffer, final int offset) {
         final int type = directBuffer.getInt(offset);
         if (type >= 0 & type <= MessageType.maxOrdinal()) {
             final MessageType messageType = MessageType.valueByOrdinal(type);
-            final DirectMessage message = messageType.create(this);
+            final AbstractDirectMessage message = messageType.create(this);
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/message/direct/DirectMessageFactory.java]: don't like that abstractXXXmessage is exposed

File: src/main/java/org/tools4j/hoverraft/state/CandidateState.java [hunk 7]
```diff
@@ -81,26 +81,17 @@ private Transition onVoteResponse(final ServerContext serverContext, final VoteR
...
+
+        if (appendRequestTerm >= currentTerm) {
             serverContext.timer().reset();
-            successful = appendToLog(serverContext, appendRequest);
-            transition = Transition.TO_FOLLOWER;
+            return Transition.TO_FOLLOWER;
         } else {
-            successful = false;
-            transition = Transition.STEADY;
+            return super.onAppendRequest(serverContext, appendRequest);
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/state/CandidateState.java:93]: not sure I like that

File: src/main/java/org/tools4j/hoverraft/state/CommandLog.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package org.tools4j.hoverraft.state;
+
+public interface CommandLog extends LogEntry, Comparable<LogEntry> {
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/state/CommandLog.java]: it is dodgy that CommandLog _is a_ LogEntry

File: src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+            throw new IllegalArgumentException("Invalid read index " + index + " for log entry with size " + commandLogEntries.size());
+        }
+    }
+
+    @Override
+    public synchronized CommandLogEntry read() {
+        if (readIndex.get() < commandLogEntries.size()) {
+            return commandLogEntries.get(readIndex.getAndIncrement());
+        }
+        throw new IllegalArgumentException("Read index " + readIndex + " has reached end of message log with size " + commandLogEntries.size());
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java]: there are no arguments. this is either IllegalStateEx or it should never be possible to get into this situation.

It should actually only occur if index=0 and log is empty. I would throw an IllegalStateException in this case

File: src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    @Override
+    public synchronized void append(final CommandLogEntry commandLogEntry) {
+        commandLogEntries.add(Objects.requireNonNull(commandLogEntry));
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java]: maybe we should clone here somehow

File: src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+
+    @Override
+    public synchronized void truncate(long index) {
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java]: method name could be clearer, e.g. `truncateIncluding(..)`

File: src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+            return CONTAINMENT.OUT;
+        } else {
+            readIndex(logEntry.index());
+            final int termAtLogEntryIndex = read().term();
+            if (logEntry.term() != termAtLogEntryIndex) {
+                return CONTAINMENT.CONFLICT;
+            } else {
+                return CONTAINMENT.IN;
+            }
+
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java]: minor: extra empty lines

File: src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+        }
+    }
+
+    @Override
+    public int compareTo(LogEntry logEntry) {
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java]: I think CommandLog should have a method
`LogEntry lastEntry()`

and not implement LogEntry and also not be comparable to LogEntry.

LogEntry itself then implements Comparable<LogEntry>

Note that `index()` should then probably become `lastIndex()` or be removed

File: src/main/java/org/tools4j/hoverraft/state/Transition.java [hunk 14]
```diff
@@ -31,21 +31,27 @@
...
+    Transition(final Role targetRole, final boolean replayEvent) {
         this.targetRole = targetRole;//nullable
+        this.replayEvent = replayEvent;
     }
 
     public final Role targetRole(final Role currentRole) {
         return this == STEADY ? currentRole : targetRole;
     }
 
+    public final boolean replayEvent() {
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/state/Transition.java:51]: nice. I like this change.

File: src/main/java/org/tools4j/hoverraft/state/direct/DirectCommandLogEntry.java [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+import org.agrona.MutableDirectBuffer;
+import org.tools4j.hoverraft.message.CommandMessage;
+import org.tools4j.hoverraft.message.direct.DirectCommandMessage;
+import org.tools4j.hoverraft.message.direct.DirectMessage;
+import org.tools4j.hoverraft.state.CommandLogEntry;
+import org.tools4j.hoverraft.state.LogEntry;
+
+import java.util.Objects;
+
+public class DirectCommandLogEntry implements CommandLogEntry, DirectMessage {
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/state/direct/DirectCommandLogEntry.java]: nice. I like this class.

File: src/main/java/org/tools4j/hoverraft/state/direct/DirectPersistentState.java [hunk 16]
```diff
@@ -62,8 +63,10 @@ public int votedFor() {
         return state.getInt(4);
     }
 
-    public MessageLog<DirectCommandMessage> commandLog() {
-        return commandLog;
+    @Override
+    public CommandLog commandLog() {
+        //FIXme
+        return null;
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/state/direct/DirectPersistentState.java]: throwing an exception is better

File: src/main/java/org/tools4j/hoverraft/transport/aeron/AeronReceiver.java [hunk 17]
```diff
@@ -27,14 +27,15 @@
...
+import org.tools4j.hoverraft.message.direct.AbstractDirectMessage;
 import org.tools4j.hoverraft.message.direct.DirectMessage;
 import org.tools4j.hoverraft.message.direct.DirectMessageFactory;
 import org.tools4j.hoverraft.transport.Receiver;
 
 import java.util.Objects;
 import java.util.function.Consumer;
 
-public class AeronReceiver implements Receiver<DirectMessage> {
+public class AeronReceiver implements Receiver<AbstractDirectMessage> {
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/transport/aeron/AeronReceiver.java]: no way --- i really don't like that AbstractDirectMessage is now exposed everywhere.

File: src/main/java/org/tools4j/hoverraft/transport/chronicle/ChronicleMessageLog.java [hunk 18]
```diff
@@ -140,4 +140,10 @@ private Method initLastWrittenIndexMethod() {
             throw new RuntimeException(e);
         }
     }
+
+    @Override
+    public long truncate(long inclusiveIndex) {
+        //FIXme implement it
+        return 0;
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/transport/chronicle/ChronicleMessageLog.java]: throw ex

File: src/main/java/org/tools4j/hoverraft/transport/embedded/InMemoryMessageLog.java [hunk 19]
```diff
@@ -81,4 +81,10 @@ public synchronized M read() {
     public synchronized void append(final M message) {
         messages.add(Objects.requireNonNull(message));
     }
+
+    @Override
+    public long truncate(long inclusiveIndex) {
+        //FIXme implement it
+        return 0;
```
Comment (terzerm) [inline src/main/java/org/tools4j/hoverraft/transport/embedded/InMemoryMessageLog.java:88]: throw ex

File: src/test/java/org/tools4j/hoverraft/server/FollowerStateTest.java [hunk 20]
```diff
@@ -68,16 +68,33 @@ public void onAppendRequest() throws Exception {
...
+
+
+        appendRequest.commandLogEntry().index(101L)
+                                       .term(term);
+
+        //make appendRequest prevLogEntry to match the end of commandLog
+        appendRequest.prevLogEntry().term(term)
+                                    .index(100);
+
+        when(persistentState.commandLog().contains(appendRequest.prevLogEntry())).thenReturn(CommandLog.CONTAINMENT.IN);
```
Comment (terzerm) [inline src/test/java/org/tools4j/hoverraft/server/FollowerStateTest.java]: extract commandLog and prefLogEntry into variable for readability

File: src/test/java/org/tools4j/hoverraft/server/VoteRequestHandlerTest.java [hunk 21]
```diff
@@ -177,13 +186,15 @@ public void onVoteRequest_wrongTerm() throws Exception {
         final VoteRequest voteRequest = DirectMessageFactory.createForWriting()
                 .voteRequest()
                 .term(badTerm)
-                .candidateId(serverId)
-                .lastLogTerm(lastLogTerm)
-                .lastLogIndex(lastLogIndex);
+                .candidateId(serverId);
+        //need LogEntry mutators to return parent object on done()
+        voteRequest.lastLogEntry().term(lastLogTerm)
+                .index(lastLogIndex);
```
Comment (terzerm) [inline src/test/java/org/tools4j/hoverraft/server/VoteRequestHandlerTest.java:194]: that's the disadvantage of the nested object structure --- the method chaining breaks down :-(

File: src/test/java/org/tools4j/hoverraft/server/VoteRequestHandlerTest.java [hunk 22]
```diff
@@ -223,13 +235,15 @@ public void onVoteRequest_invalidCandidate_badLastLogIndex() throws Exception {
...
+                .candidateId(serverId);
+        //need LogEntry mutators to return parent object on done()
+        voteRequest.lastLogEntry().term(lastLogTerm)
+                .index(badLastLogIndex);
+
         when(serverContext.connections().serverSender(serverId)).thenReturn(sender);
         when(persistentState.votedFor()).thenReturn(PersistentState.NOT_VOTED_YET);
-        when(persistentState.lastLogTerm()).thenReturn(lastLogTerm);
-        when(persistentState.lastLogIndex()).thenReturn(lastLogIndex);
+        when(persistentState.commandLog().compareTo(voteRequest.lastLogEntry())).thenReturn(1);
```
Comment (terzerm) [inline src/test/java/org/tools4j/hoverraft/server/VoteRequestHandlerTest.java]: extract commandLog and lastLogEntry into variable for readability

File: src/test/java/org/tools4j/hoverraft/state/InMemoryCommandLogTest.java [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+        when(commandLogEntry1.index()).thenReturn(1L);
+
+
+        InMemoryCommandLog commandLog = new InMemoryCommandLog();
+        commandLog.append(commandLogEntry1);
+
+        assertThat(commandLog.size()).isEqualTo(1);
+
+
+    }
```
Comment (terzerm) [inline src/test/java/org/tools4j/hoverraft/state/InMemoryCommandLogTest.java:26]: poor test coverage

**Review by terzerm — CHANGES_REQUESTED** [review]
Must fixes in my opinion:
- exposure of AbstractDirectMessage
- CommandMessageLog implementing LogEntry --- it should be _has a_ instead of _is a_ relation

File: src/main/java/org/tools4j/hoverraft/message/MessageType.java (hunk 3 above)
Comment (anton-anufriev) [reply src/main/java/org/tools4j/hoverraft/message/MessageType.java]: Yeah, I did think about that and left it for discussion. Fixed now.

File: src/main/java/org/tools4j/hoverraft/message/direct/DirectAppendRequest.java (hunk 5 above)
Comment (anton-anufriev) [reply src/main/java/org/tools4j/hoverraft/message/direct/DirectAppendRequest.java:121]: fixed

File: src/main/java/org/tools4j/hoverraft/message/direct/DirectMessageFactory.java (hunk 6 above)
Comment (anton-anufriev) [reply src/main/java/org/tools4j/hoverraft/message/direct/DirectMessageFactory.java]: fixed

File: src/main/java/org/tools4j/hoverraft/state/CommandLog.java (hunk 8 above)
Comment (anton-anufriev) [reply src/main/java/org/tools4j/hoverraft/state/CommandLog.java]: Agree, fixed.

File: src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java (hunk 9 above)
Comment (anton-anufriev) [reply src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java]: Copied from your InMemoryMessageLog. Agree that IllegalStateEx is better.

Comment (anton-anufriev) [reply src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java]: Fixed in both

File: src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java (hunk 10 above)
Comment (anton-anufriev) [reply src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java]: I think it is better that InMemoryFactory with InMemoryMessages

File: src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java (hunk 11 above)
Comment (anton-anufriev) [reply src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java]: done

File: src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java (hunk 12 above)
Comment (anton-anufriev) [reply src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java]: easy

File: src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java (hunk 13 above)
Comment (anton-anufriev) [reply src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java]: Yes, I have done it.

File: src/main/java/org/tools4j/hoverraft/state/direct/DirectPersistentState.java (hunk 16 above)
Comment (anton-anufriev) [reply src/main/java/org/tools4j/hoverraft/state/direct/DirectPersistentState.java]: nice! done

File: src/main/java/org/tools4j/hoverraft/transport/aeron/AeronReceiver.java (hunk 17 above)
Comment (anton-anufriev) [reply src/main/java/org/tools4j/hoverraft/transport/aeron/AeronReceiver.java]: it was a temporary change. Fixed

File: src/test/java/org/tools4j/hoverraft/server/FollowerStateTest.java (hunk 20 above)
Comment (anton-anufriev) [reply src/test/java/org/tools4j/hoverraft/server/FollowerStateTest.java]: done

File: src/test/java/org/tools4j/hoverraft/server/VoteRequestHandlerTest.java (hunk 22 above)
Comment (anton-anufriev) [reply src/test/java/org/tools4j/hoverraft/server/VoteRequestHandlerTest.java]: done

File: src/main/java/org/tools4j/hoverraft/state/CandidateState.java (hunk 7 above)
Comment (anton-anufriev) [reply src/main/java/org/tools4j/hoverraft/state/CandidateState.java:93]: why?

**Pull request merged**